### PR TITLE
8261949: fileStream::readln returns incorrect line string

### PR DESCRIPTION
**Pull request by kelthuzadx**
When the last line does not contain a NEWLINE character, fileStream::readln would read
truncated line string:

$ cat file_content:
AA
BB
CC<EOF>

fileStream::readln result:
"AA"
"BB"
"C"

This patch address this problem, it works for Posix and Windows since the last character
of these systems is always '\n'.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261949](https://bugs.openjdk.java.net/browse/JDK-8261949): fileStream::readln returns incorrect line string

### Reviewers
 * [Daniel D. Daugherty](https://openjdk.java.net/census#dcubed) (@dcubed-ojdk - **Reviewer**)
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2626/head:pull/2626`
`$ git checkout pull/2626`

### CODE REVIEW

File: src/hotspot/share/utilities/ostream.cpp [hunk 1]
```diff
@@ -589,8 +589,11 @@ char* fileStream::readln(char *data, int count ) {
   char * ret = NULL;
   if (_file != NULL) {
     ret = ::fgets(data, count, _file);
-    //Get rid of annoying \n char
-    data[::strlen(data)-1] = '\0';
+    // Get rid of annoying \n char only if it presents, it works for Posix
+    // and Windows since the last character of these systems is always '\n'
```
Comment (dcubed-ojdk) [inline src/hotspot/share/utilities/ostream.cpp:593]: s/of these/on these/
Please end the comment with a period.

File: src/hotspot/share/utilities/ostream.cpp [hunk 2]
```diff
@@ -589,8 +589,11 @@ char* fileStream::readln(char *data, int count ) {
...
+    if (data[::strlen(data)-1] == '\n') {
+      data[::strlen(data)-1] = '\0';
```
Comment (dcubed-ojdk) [inline src/hotspot/share/utilities/ostream.cpp:595]: Perhaps:
```
    size_t last_char = ::strlen(data) - 1;
    if (last_char >= 0 && data[last_char] == '\n') {
      data[last_char] = '\0';
```

**Review by dcubed-ojdk — CHANGES_REQUESTED** [review]

File: src/hotspot/share/utilities/ostream.cpp (hunk 1 above)
Comment (dholmes-ora) [reply src/hotspot/share/utilities/ostream.cpp:593]: I suggest simply:

// Get rid of \n char if it is present

File: src/hotspot/share/utilities/ostream.cpp (hunk 2 above)
Comment (dholmes-ora) [reply src/hotspot/share/utilities/ostream.cpp:595]: @dcubed-ojdk size_t is unsigned so by definition always >= 0. :) I suggest:
`size_t len = ::strlen(data);`
`if (len > 0 && data[len - 1] == '\n') {`
`  data[len - 1] = '\0';`
`}`

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

File: src/hotspot/share/utilities/ostream.cpp (hunk 1 above)
Comment (kelthuzadx) [reply src/hotspot/share/utilities/ostream.cpp]: Changed.

File: src/hotspot/share/utilities/ostream.cpp (hunk 2 above)
Comment (kelthuzadx) [reply src/hotspot/share/utilities/ostream.cpp]: Make sense. Changed.

**Review by dcubed-ojdk — APPROVED** [review]
I like David's suggestion better than mine.
Thumbs up.

**Review by dholmes-ora — APPROVED** [review]
Looks good!

Thanks,
David

**Comment by kelthuzadx** [comment]
Thanks @dcubed-ojdk  @dholmes-ora  for reviews! 

Would you be able to sponsor this patch? I'm not a committer so that I can not push this commit directly.

**Comment by kelthuzadx** [comment]
@DamonFool I'm grateful for your sponsorship.

**Pull request closed without merging**